Remove duplicate CiliumNode watcher

aanm opened the pull request:
In certain cases the Cilium Operator could end up with two CiliumNode
k8s watchers. This PR re-writes the code in a way that a single
watcher will be used for all cases.

aanm committed:
operator: merge duplicate k8s watchers

In certain cases the Cilium Operator could end up with two CiliumNode
k8s watchers. This commit re-writes the code in a way that a single
watcher will be used for all cases.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
operator: refator code from k8s_node.go to cilium_node.go

The code present in k8s_node was not specific to a k8s node. Thus, this
commit moved the code into a better appropriate file, cilium_node.go

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
operator: remove unused 'nodes-gc-interval' flag

This flag is no longer used in the code so we can safely remove it.

Fixes: 3aacf41d784e ("add capability to disable CNP NodeStatus updates")
Signed-off-by: André Martins <andre@cilium.io>

jrajahalme commented on operator/cilium_node.go:
```diff
@@ -20,16 +26,240 @@ import (
+	withKVStore           bool
+}
+
+// ciliumNodeWatcherCfgBuilder is used to generate a ciliumNodeWatcherCfg
```
I would think that a `....Builder` is a function type. Maybe better name this simply as `ciliumNodeWatcherConfig`??

aanm replied:
@jrajahalme how would we distinguish between this struct and the other one from line 43?

ldelossa replied:
I don't think it's unreasonable to have a structure called "...Builder" with a method set that builds something. In this case, a config.

jrajahalme commented on operator/cilium_node.go:
```diff
@@ -20,16 +26,240 @@ import (
+}
+
+// NewCiliumNodeWatcherCfgBuilder returns a new ciliumNodeWatcherCfgBuilder.
+func NewCiliumNodeWatcherCfgBuilder() *ciliumNodeWatcherCfgBuilder {
```
Same comment, please drop the `Builder` suffix :-)

aanm replied:
Same comment above.

jrajahalme commented on operator/cilium_node.go:
```diff
@@ -20,16 +26,240 @@ import (
+			}
+		} else {
+			// nodeManager not nil thus the events will be processed by
+			// ciliumNodeKVStore and the nodeManager
```
There is a lot of code duplication here, would it be better to only have one set of resource event handler funcs that would work both with nil and non-nil nodeManager?

aanm replied:
There is, I was worried on optimizing the code over readbility. Please see the commit _operator: simplify duplicated code in the event handling_ after this one where I've simplified the code a little bit.

jrajahalme commented on operator/cilium_node.go:
```diff
@@ -20,16 +26,240 @@ import (
+		} else {
+			// nodeManager not nil but the KVStore is nil thus the events
+			// will only be relevant to the nodeManager.
+			cfg.resourceEventHandler = cache.ResourceEventHandlerFuncs{
```
Do we really want 3 sets of these funcs with repeated conversion and error handling?

jrajahalme reviewed (approved):
Looks OK on the top level, but I was not happy with 3 sets of handler funcs which all repeat the same conversion and error handling.

joestringer reviewed (approved):
LGTM from docs perspective. Looks like Jarno already reviewed the core functionality here, so I skipped that for now. Let me know if you would like additional review.

aanm committed:
operator: simplify duplicated code in the event handling

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
operator: use workqueue for event handling of CiliumNodes

As this code suffers a major rewrite, this commit introduces the right
way of handling Kubernetes events, with a workqueue.

Signed-off-by: André Martins <andre@cilium.io>

joestringer commented on operator/cilium_node.go:
```diff
@@ -1,112 +1,486 @@
+	return cfgBuilder
+}
+
+// Build builds and if WithKVStoreSync was called it will start a SharedStore.
```
Build builds what? :joy: 
```suggestion
// Build builds the configuration and if WithKVStoreSync was called it will start a SharedStore.
```

joestringer commented on operator/main.go:
```diff
@@ -376,8 +376,9 @@ func onOperatorStartLeading(ctx context.Context) {
 	var (
-		nodeManager *allocator.NodeEventHandler
-		err         error
+		cnWatcherCfgBuilder = NewCiliumNodeWatcherCfgBuilder()
```
nit: It seems error-prone to me to initialize this object partially here, partially a hundred lines down, partially another hundred lines down, then finish initialization and start goroutines etc just below that. Can we just track a few booleans and initialize & start it all at once?

More specifically, I think it's a concerning sign if the builder pattern (ie `foo.WithBar()`) is used all by itself and not as part of a block that does:
```
foo :=NewFoo().WithAlice()
if bar {
    foo = foo.WithBar()
}
if bob {
    foo = foo.WithBob()
}
```
The reasoning is similar; basically it's hard for the reader to understand that the code 100 lines down can safely modify the structure in a way that an initializer function like this would do. How do we know that some other logic isn't already consuming the structure and we're just trashing memory others are accessing? Well, we can dig around and find all the references to the structure.... or in the first place we just initialize once in a single block, then it's easier to tell if it's doing the right thing.

joestringer commented on operator/main.go:
```diff
@@ -398,42 +399,15 @@ func onOperatorStartLeading(ctx context.Context) {
-		}
-	}
-
 	if kvstoreEnabled() {
```
If the nodewatcher is no longer run above this block, could we miss handling of some events correctly?

aanm replied:
The are independent, the nodewatcher is not used in the `if kvstoreEnabled() {` branch.

joestringer replied:
Looking more deeply into the `runNodeWatcher()` function, it just kicks off two async goroutines to GC CNP/CCNP node status anyway so the previous code wasn't providing any guarantees about when it would run. So should be fine from that perspective.

joestringer commented on operator/main.go:
```diff
@@ -509,14 +483,41 @@ func onOperatorStartLeading(ctx context.Context) {
+		// knows all podCIDRs that are currently set in the cluster, that
+		// it can allocate podCIDRs for the nodes that don't have a podCIDR
+		// set.
+		nodeManager.Resync(context.Background(), time.Time{})
```
Is there no context for the operator? :thinking: we should add one (maybe just file an issue, could be a good-first issue maybe?).

ldelossa commented on operator/cilium_node.go:
```diff
@@ -20,16 +26,240 @@ import (
+// NewCiliumNodeWatcherCfgBuilder returns a new ciliumNodeWatcherCfgBuilder.
+func NewCiliumNodeWatcherCfgBuilder() *ciliumNodeWatcherCfgBuilder {
+	var cfgBuilt atomic.Value
+	cfgBuilt.Store(false)
```
You can probably drop the atomic.Value for a simple integer type and call atomic.AddInt32/atomic.LoadInt32. The API maybe a bit nicer to deal with.

ldelossa commented on operator/cilium_node.go:
```diff
@@ -20,16 +26,240 @@ import (
+// WithNodeManager sets the ciliumNodeWatcherCfgBuilder with a nodeManager that
+// will be used in the CiliumNode watch events.
+func (cfgBuilder *ciliumNodeWatcherCfgBuilder) WithNodeManager(nodeManager allocator.NodeEventHandler) *ciliumNodeWatcherCfgBuilder {
+	// We shouldn't change modify cfgBuilder.nodeManager after a
```
```suggestion
	// We shouldn't modify cfgBuilder.nodeManager after a
```

joestringer commented on operator/main.go:
```diff
@@ -509,14 +483,41 @@ func onOperatorStartLeading(ctx context.Context) {
+	if err != nil {
+		log.WithError(err).Fatal("Unable to generate configuration for CiliumNode watcher")
+	}
+	if err := startSynchronizingCiliumNodes(cnWatcherCfg); err != nil {
```
I'm confused why the builder needs to exist here, why couldn't we just do something like:
```suggestion
	cnSync, err := NewCiliumNodeSynchronizer(cnWatcherCfg)
	if err != nil {
		log.WithError(err).Fatal("Unable to generate configuration for CiliumNode watcher")
	}
	if err := cnSync.startSynchronizingCiliumNodes(); err != nil {
```

ldelossa commented on operator/cilium_node.go:
```diff
@@ -20,16 +26,240 @@ import (
+	return cfgBuilder
+}
+
+// Build builds and if WithKVStoreSync was called it will start a SharedStore.
```
```suggestion
// Build builds the config and if WithKVStoreSync was called it will start a SharedStore.
```

joestringer commented on operator/cilium_node.go:
```diff
@@ -1,112 +1,486 @@
-		(*nodeManager).Delete(name)
+// processNextWorkItem process all events from the workqueue. See
+// workqueue.RateLimitingInterface documentation to understand why certain
+// methods are executed in certain conditions.
```
What does this comment mean? I looked up https://pkg.go.dev/k8s.io/client-go/util/workqueue#RateLimitingInterface and I'm now more confused which conditions or methods this comment is trying to explain.

joestringer commented on operator/cilium_node.go:
```diff
@@ -1,112 +1,486 @@
+		// there isn't any event handler set for CiliumNodes events.
+		if cfg.syncHandler != nil {
+			for processNextWorkItem(cfg.queue, cfg.syncHandler) {
+			}
```
I feel like this code is trying to trick me. At face value, the actual execution for each item that's processed is "do nothing", but actually the "do something" is a byproduct of the iterator. This code would be easier to follow if it were just opencoding the implementation like `for key, quit := cfg.queue.Get(); !quit { doStuff() }`.

joestringer replied:
(On second read, I see that the actual execution is running `processNextWorkItem()` so it's clear that it will do something, but it's the empty braces that mislead me here.)

ldelossa commented on operator/cilium_node.go:
```diff
@@ -20,16 +26,240 @@ import (
+		// for KVStore.
+		if cfgBuilder.nodeManager == nil {
+			cfg.resourceEventHandler = cache.ResourceEventHandlerFuncs{
+				AddFunc: func(obj interface{}) {
```
I think for readability take, pull out these function literals to somewhere outside the if/else block and give them semantic meaning. Such as "AddFuncNilNodeManager"

joestringer commented on operator/cilium_node.go:
```diff
@@ -1,112 +1,486 @@
+// workqueue.RateLimitingInterface documentation to understand why certain
+// methods are executed in certain conditions.
+func processNextWorkItem(queue workqueue.RateLimitingInterface, syncHandler func(key string) error) bool {
+	key, quit := queue.Get()
+	if quit {
+		return false
+	}
```
When does this quit? Is there any possibility that this will somehow stop processing events but the overall operator will continue running just without the functionality that's supposed to be handled in this workqueue?

(Maybe put another way, when this quits, should we log an Info log or something that the operator will stop processing these work items?)

aanm replied:
This should only quit when the operator terminates.

ldelossa commented on operator/cilium_node.go:
```diff
@@ -136,40 +135,77 @@ func (cfgBuilder *ciliumNodeWatcherCfgBuilder) Build() (cfg *ciliumNodeWatcherCf
+			if !exists && errors.IsNotFound(err) {
+				if cfgBuilder.kvStoreSync {
+					ciliumNodeKVStore.DeleteLocalKey(context.TODO(), &ciliumNodeName{name: name})
+				}
```
Do we want to handle those context.TODO()'s before merging?

joestringer reviewed (commented):
Some minor comments from a deeper review (mostly focused on the first and last commits) but I think the implementation looks correct.

Mainly I found the Builder structure confusing. I'm used to reading Cilium code that says something like, 

```
func NewFooer(cfg Config) *fooer { return &fooer{....}
func (f *fooer) WithBar(b Bar) *fooer { f.Bar = b; return f }
func (f *fooer) Run() {
    if f.FooEnabled {
        doFoo()
    }
    if f.BarEnabled {
        doBar()
    }
}
```

However, here there's no Fooer. There's a config builder (which does stuff like syncing state with the kvstore, which is a bit surprising), then that returns a config which is passed to a static function that operates against that config.

I'm sure there's reasons for this approach, and maybe some of it is that k8s code looks more like this so I'm just out of my depth a bit. But I at least found it difficult to follow.

aanm committed:
operator: use context of leader election for background operations

Signed-off-by: André Martins <andre@cilium.io>

aanm reviewed (commented):
Thanks for the review

ldelossa reviewed (approved):
My feedback was handled. Approving contingent on other's reviews.

joestringer commented on operator/cilium_node.go:
```diff
@@ -1,112 +1,428 @@
+	if quit {
+		return false
+	}
+	// Mark it as Done
```
nit: I suspect that these were added in response to my earlier feedback, but a couple of these comments are superfluous, the code `queue.Get()` clearly gets the next event from the queue. Similarly the `queue.Done()` call marks it as done, that's clear from the code. If that's how the workqueue interface works, we don't have a choice and the comments don't help explain why, they just duplicate the explanation from the code in different words. I think it's fine to drop those comments.

```suggestion
	key, quit := queue.Get()
	if quit {
		return false
	}
```

joestringer reviewed (approved):
LGTM. I'm sure the more I look, the more I could find ways we can incrementally tidy up the code but I think it'll just become bikeshedding at that point. :ship:

jrajahalme reviewed (approved):
LGTM

christarazi commented on operator/cilium_node.go:
```diff
@@ -164,13 +203,38 @@ func startSynchronizingCiliumNodes(nodeManager allocator.NodeEventHandler, withK
+		return true
+	}
+
+	return true
```
I recall @ldelossa mentioning some nuance where in some case (forgot what it was) that we need to make some extra call to `queue.Add()` or similar. It that addressed here? Louis, could you remind me what that was and validate whether this code already addressed it?

aanm replied:
Nice call @christarazi I've added that missing piece.

christarazi reviewed (commented):
Changes LGTM, one question below regarding workqueue handling.
